lib/collection/src/shards/replica_set/clock_set.rs

```diff
@@ -0,0 +1,94 @@
+use std::cmp;
+use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
+use std::sync::Arc;
+
+#[derive(Clone, Debug, Default)]
+pub struct ClockSet {
+    clocks: Vec<Arc<Clock>>,
+}
+
+impl ClockSet {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Get the first available clock from the set, or create a new one.
+    pub fn get_clock(&mut self) -> ClockGuard {
+        for (id, clock) in self.clocks.iter().enumerate() {
+            if clock.lock() {
+                return ClockGuard::new(id, clock.clone());
+            }
+        }
+
+        let id = self.clocks.len();
+        let clock = Arc::new(Clock::new_locked());
+
+        self.clocks.push(clock.clone());
+
+        ClockGuard::new(id, clock)
+    }
+}
+
+#[derive(Debug)]
+pub struct ClockGuard {
+    id: usize,
+    clock: Arc<Clock>,
+}
+
+impl ClockGuard {
+    fn new(id: usize, clock: Arc<Clock>) -> Self {
+        Self { id, clock }
+    }
+
+    pub fn id(&self) -> usize {
+        self.id
+    }
+
+    #[must_use = "new clock value must be used"]
+    pub fn tick_once(&mut self) -> u64 {
+        self.clock.tick_once()
+    }
+
+    pub fn advance_to(&mut self, new_tick: u64) -> u64 {
+        self.clock.advance_to(new_tick)
+    }
+}
+
+impl Drop for ClockGuard {
+    fn drop(&mut self) {
+        self.clock.release();
+    }
+}
+
+#[derive(Debug)]
+struct Clock {
+    clock: AtomicU64,
+    available: AtomicBool,
+}
+
+impl Clock {
+    fn new_locked() -> Self {
+        Self {
+            clock: AtomicU64::new(0),
+            available: AtomicBool::new(false),
+        }
+    }
+
+    #[must_use = "new clock value must be used"]
+    fn tick_once(&self) -> u64 {
+        self.clock.fetch_add(1, Ordering::Relaxed)
+    }
+
+    fn advance_to(&self, new_tick: u64) -> u64 {
+        let current_tick = self.clock.fetch_max(new_tick + 1, Ordering::Relaxed);
+        cmp::max(current_tick, new_tick + 1)
+    }
```

Comment:
Ok, so, after discussion this morning, I've updated `Clock` so that it will start ticking from `0` instead of `1`... but, can you guys please help me analyze if I've updated `advance_to` correctly? 🥲🙏

Initially
- `tick_once` was `fetch_add(1) + 1`
- and `advance_to` was `fetch_max(new_tick)`

But now
- `tick_once` does `fetch_add(1)`
  - so it _does_ increment, but returns the value _before_ increment
  - so if `Clock` starts from `0`, then `tick_once` will return `0` and store `1`
  - then the next `tick_once` will return `1` and store `2`
  - etc...
- so, with that in mind, _I think_, it makes sense to change `advance_to` to `fetch_max(new_tick + 1)`
  - e.g., we restarted a node, the clock is `0`, we send first request and receive response with clock `10`
  - then we have to do `fetch_max(10 + 1)`
  - so that _the next_ `tick_once` will return `11` (and store `12`)
  - this, _kinda_, depends on how exactly we calculate the clock in update operation response
  - e.g., we could return "current version" or "current version + 1"
  - but it seems easier to return "current version"
  - and then we need to do `fetch_max(new_tick + 1)`
  - _I think_

Reply:
@generall @timvisee